Fix/1443 remove js runtime from threshold calculation

oleiade opened the pull request:
This Pull Request does the following:
* brings back the reverted threshold parsing.
* adds support for the `--no-thresholds` option of the `run` command
* introduces a `Thresholds.Parse` method. Calls it as part of the `run` command
* modifies tests in `thresholds_test.go` which assumed `UnmarshalJSON` would parse thresholds

This PR acts as a tentative to address the minimum scope necessary to merge Thresholds parsing safely to master. It will need to lead to further refactors and improvements both in k6 itself, and in our internal infrastructure.

Reviewers can safely ignore the merge commit, as it was already approved in https://github.com/grafana/k6/pull/2251. Reviewers should focus on the last commit only.

MStoykov commented on stats/thresholds_test.go:
```diff
@@ -56,9 +54,8 @@ func TestNewThreshold_InvalidThresholdConditionExpression(t *testing.T) {
-	gotThreshold, err := newThreshold(src, abortOnFail, gracePeriod)
+	gotThreshold := newThreshold(src, abortOnFail, gracePeriod)
 
-	assert.Error(t, err, "instantiating a threshold with an invalid expression should fail")
```
Can you instead of removing this, call Parse and check the error from it?

oleiade replied:
To clarify, do you mean calling `Parse` in the test here, or calling it in `UnmarshalJSON`?

PS: calling it inside the test would be equivalent to test `Parse` itself. But I want to make sure :)

MStoykov replied:
I meant calling it here on its own, so you can just get the same error you got before. 

This also is true for all other places – just everywhere call `Parse` after the thresholds are created and check the error

oleiade replied:
Alright, will do 👍🏻

oleiade replied:
I actually had forgotten to remove `parseThresholdExpression` call from `newThreshold`. It's now done. You did in your patch too. It makes sense.

Following that, I don't think it makes much sense to call `Parse` in the `newThreshold` (singular) tests anymore, as it would only test the input we give in the test is parsable, and would not assert anything about `newThreshold` itself anymore. Would you agree?

MStoykov replied:
On a similar note the current test failures are because you don't Parse it in some cases where that is needed as it will then be used. 

Given that this in practice just restores to what hte test were before I feel that is much better either way.

oleiade replied:
Indeed, I caught that in the process. Thanks for pointing it out anyway 😸 I've added parsing steps to the tests.

MStoykov commented on stats/thresholds.go:
```diff
@@ -44,18 +44,18 @@ type Threshold struct {
 	if err != nil {
-		return nil, err
+		return nil
 	}
```
You need to remove this lines otherwise this will return a nil an then you will get nullPointerException

oleiade replied:
Good catch Indeed 👀  I actually ran into the problem while writing test for `Parse` 👍🏻

MStoykov reviewed (commented):
Apart from: 
1. the one thing that leads to nil pointer exceptions that I did actually fix in my patch
2.  the tests not calling `Parse` -which probably would've caught `1`

I think this  okay :+1:

oleiade commented:
As no one had committed to a review yet, I've force-pushed. Sorry if that caused any disturbance. I have removed the dependency of `newThreshold` (**singular**) to `parseThresholdExpression`. I have added tests for the `Thresholds.Parse` (**plural**) method.

MStoykov commented on cmd/run.go:
```diff
@@ -125,6 +125,16 @@ a commandline interface for interacting with it.`,
 				return err
 			}
 
+			// Parse the thresholds, only if the --no-threshold flag is not set.
+			if !runtimeOptions.NoThresholds.Bool {
+				for _, thresholds := range conf.Options.Thresholds {
+					err := thresholds.Parse()
+					if err != nil {
+						return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+					}
+				}
+			}
+
```
Unfortunately I think we need to copy this to `cmd/cloud.go` and `cmd/archive.go` :thinking: Looking at the code the alternative is to add it in `getConsolidatedConfig`, which arguably will make this more testable :thinking:

oleiade replied:
I've added the check to `cmd/cloud.go` and `cmd/archive.go`. I've added tests to verify the behavior is respected in the context of the `run` command. Looking into doing the same for `cloud` and `archive`. 

I chose not to put the code in `getConsolidatedConfig` as the actual check is really a side effect, one that does not have any impact on the config struct `getConsolidateConfig` produces (as far as I can tell). Moreover, I do think that validation before running will justify refactoring this somehow in the close future anyway.

olegbespalov commented on cmd/archive.go:
```diff
@@ -75,6 +77,18 @@ An archive is a fully self-contained test run, and can be executed identically e
+			// If parsing the threshold expressions failed, consider it as an
+			// invalid configuration error.
+			if !runtimeOptions.NoThresholds.Bool {
+				for _, thresholds := range conf.Options.Thresholds {
```
What if we move this loop to the `conf.options.ValidateTreshholds` ? :thinking:

MStoykov replied:
I think that is probably a small enough change to be done, but given that we are trying to get it merged before the release - I vote for leaving this for the next PR in the next release that will either way add a bunch of Validation.

olegbespalov commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	// Otherwise, attempt to parse a percentile expression
+	if strings.HasPrefix(input, tokenPercentile+"(") && strings.HasSuffix(input, ")") {
```
If we change the logical condition to, we can benefit from the early return with the error that strictly follows from the check, like:
```
if !strings.HasPrefix(input, tokenPercentile+"(") || !strings.HasSuffix(input, ")") {
     return "", null.Float{}, fmt.Errorf("failed parsing method from expression")
}
```

And this change also reduces the nesting of the ifs, which improves readability.

WDYT?

MStoykov reviewed (approved):
LGTM - I am going to make some more manual tests but hopefully we don't need to change anything to get this in v0.37.0 and make other non necessary fixes for v0.38.0

olegbespalov reviewed (approved):
Okay, keeping my comments for the following PRs :+1:

oleiade commented:
Thanks everyone, for your input 🎉 
Merging 🦖
